### PR TITLE
[ALLUXIO-3183] Support properties with NULL values in Web UI and CLI

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-3183

Major change in this PR:
- Make `Configuration.PROPERTIES` the single source of truth for all types of property keys/vals, including property keys with no values set.
- Make CLI and WebUI show properties with no value set.

Reason to have `Configuration.PROPERTIES` the single source of truth:

There are 4 types of keys in in 1.8-snapshot branch:
- TypeA: Defined in PropertyKeys and with default value
- TypeB: Defined in PropertyKeys and with no value set
- TypeC: Defined in PropertyKeys and value set in system/site prop
- TypeD: Not defined in PropertyKeys (e.g., derived from template or UFS extension) and value set in system/site prop

Before this PR,  `PROPERTIES` keeps track of type C and D. There is confusion regarding the source of truth for all possible properties
- in `getNestedProperties`, `checkTieredLocality` and etc, PROPERTIES (C + D) is considered as the entire key-value pool
- in `containsKey`, `toMap`, `toRawMap`, PROPERTIES (C +D)  + type A  is considered

I think it will be simpler to have `PROPERTIES` rather than iterating with multiple parties (A B C D). Originally `PROPERTIES` is a concurrent hash map which does not support null value. This PR works this around by changing the value type from `String` to `Optional<String>`.

### CODE REVIEW

**Comment by aaudiber** [comment]
I like the idea of having a single source of truth for configuration. My only concern is losing support for UFS extensions which rely on their own default property keys. To achieve both, could we add another layer of abstraction by moving PROPERTIES to its own class? Then we can enforce that all usages support every type of property. Something like this:

```java
public class AlluxioProperties {
  /** Map of user-supplied properties. */
  private static final ConcurrentHashMap<String, String> PROPERTIES = new ConcurrentHashMap<>();

  public Optional<String> get(String key) {
    if (PROPERTIES.containsKey(property)) {
      return Optional.of(PROPERTIES.get(property));
    }
    if (!PropertyKey.isValid(property)) {
      throw new RuntimeException("Invalid key encountered: " + property);
    }
    return Optional.ofNullable(PropertyKey.fromString(property).getDefaultValue());
  }

  // Implement other methods that we currently call on Configuration#PROPERTIES.
}
```

**Comment by LuQQiu** [comment]
That's a good fix to 1.7, adding a optional<string> we are already able to support null values! I just wondering do we have a separate UFS configuration that stores and manages all the UFS related properties?

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+      return Source.DEFAULT;
+    }
+    return Source.UNKNOWN;
+  }
+
+  /**
+   * Merges the current configuration properties with alternate properties. A property from the new
+   * configuration wins if it also appears in the current configuration.
+   *
+   * @param properties The source {@link Properties} to be merged
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: The -> the 
here and on the line below to be consistent within this file

File: core/common/src/main/java/alluxio/conf/Source.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.conf;
+
+/**
+ * The source of a configuration property.
+ */
+public enum Source {
+  DEFAULT(0),
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/conf/Source.java]: I'd prefer ordering by priority over alphabetically

File: core/common/src/main/java/alluxio/Configuration.java [hunk 3]
```diff
@@ -64,99 +65,71 @@
...
-  /** Map of properties. */
-  private static final ConcurrentHashMap<String, String> PROPERTIES = new ConcurrentHashMap<>();
-  /** Map of property sources. */
-  private static final ConcurrentHashMap<PropertyKey, Source> SOURCES = new ConcurrentHashMap<>();
+  /** Source of the truth of all property values (default or customized). */
+  private static AlluxioProperties sProperties;
+  /** Path of site properties file. */
   private static String sSitePropertyFile;
+  /** Map from source to properties. */
+  private static Map<Source, Map<?, ?>> sSources = Maps.newHashMap();
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/Configuration.java]: It's a little weird that this hangs around here even though it isn't a source of truth, perhaps we could push the source management logic into `AlluxioProperties`? That would fit with the separation we discussed where `AlluxioProperties` holds the config values while `Configuration` builds extra functionality (validation, type conversion) on top of `AlluxioProperties`

File: core/common/src/main/java/alluxio/Configuration.java [hunk 4]
```diff
@@ -64,99 +65,71 @@
...
         siteProps = ConfigurationUtils.loadPropertiesFromResource(Constants.SITE_PROPERTIES);
       }
-      if (siteProps != null) {
-        // Update site properties and system properties in order
-        merge(siteProps, Source.SITE_PROPERTY);
-        merge(systemProps, Source.SYSTEM_PROPERTY);
-      }
     }
 
+    // Step3: re-create the configuration instance after figuring out the site properties file.
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/Configuration.java]: fixed:
- in test mode: only step1
- in non-test mode: still step 1,2,3 (we can not skip step1)

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+   *
+   * @param key key to put
+   * @param value value to put
+   */
+  public void put(String key, String value) {
+    mUserProps.put(key, Optional.ofNullable(value));
+  }
+
+  /**
+   * Remove the value set to key.
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: done

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+   *
+   * @param key key to put
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: done

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    if (mUserProps.containsKey(key)) {
+      return mUserProps.get(key).isPresent();
+    }
+    return PropertyKey.isValid(key) && (PropertyKey.fromString(key).getDefaultValue() != null);
+  }
+
+  /**
+   * @return the entry set of all Alluxio property key and value pairs (value can be null)
+   */
+  public Set<Map.Entry<String, String>> entrySet() {
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: done

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+                .map(key -> Maps.immutableEntry(key.getName(), key.getDefaultValue())),
+            mUserProps.entrySet().stream()
+                .map(entry -> Maps.immutableEntry(entry.getKey(), entry.getValue().orElse(null))))
+            .collect(toSet());
+
+    return entrySet;
+  }
+
+  /**
+   * @return the entry set of all Alluxio property key and value pairs (value can be null)
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: done

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java (hunk 1 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: fixed

File: core/common/src/main/java/alluxio/conf/Source.java (hunk 2 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/Source.java]: done

File: core/common/src/main/java/alluxio/Configuration.java (hunk 3 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/Configuration.java]: Removed from this class to keep this Configuration class involves minimally in different sources.

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+      return mUserProps.get(key).orElse(null);
+    }
+    // fromString will check if the key is valid or throw RTE
+    return PropertyKey.fromString(key).getDefaultValue();
+  }
+
+  /**
+   * Clears all existing user-specified properties.
+   */
+  public void clear() {
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: now it is needed after my latest change

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+    }
+  }
+
+  /**
+   * Sets the source for a given key.
+   *
+   * @param key property key
+   * @param source the source
+   */
+  public void setSource(String key, Source source) {
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: removed

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  /**
+   * Merges the current configuration properties with alternate properties. A property from the new
+   * configuration wins if it also appears in the current configuration.
+   *
+   * @param properties The source {@link Properties} to be merged
+   * @param source The source of the the properties (e.g., system property, default and etc)
+   */
+  public void merge(Map<?, ?> properties, Source source) {
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: now it is used in `Configuration.merge`

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+public class AlluxioProperties {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioProperties.class);
+
+  /**
+   * Map of user-specified properties. When key is mapped to Optional.empty(), it indicates no
+   * value is set for this key. Note that, ConcurrentHashMap requires not null for key and value.
+   */
+  private final ConcurrentHashMap<String, Optional<String>> mUserProps = new ConcurrentHashMap<>();
+  /** Map of property sources. */
+  private final ConcurrentHashMap<String, Source> mSources = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: done

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+ */
+@NotThreadSafe
+public class AlluxioProperties {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioProperties.class);
+
+  /**
+   * Map of user-specified properties. When key is mapped to Optional.empty(), it indicates no
+   * value is set for this key. Note that, ConcurrentHashMap requires not null for key and value.
+   */
+  private final ConcurrentHashMap<String, Optional<String>> mUserProps = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: done

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
@aaudiber I addressed the comments and fix the issue on registering existing `PropertyKey` which we talked offline. PTAL

File: core/common/src/test/java/alluxio/conf/AlluxioPropertiesTest.java [hunk 14]
```diff
@@ -39,6 +40,12 @@
       "value").build();
   private PropertyKey mKeyWithoutValue = new PropertyKey.Builder("key.without.value").build();
 
+  @After
+  public void after() {
+    PropertyKey.unregister(mKeyWithValue);
```
Comment (aaudiber) [inline core/common/src/test/java/alluxio/conf/AlluxioPropertiesTest.java]: for symmetry we could build these in `@Before`

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 15]
```diff
@@ -3586,28 +3591,37 @@ private PropertyKey(String name) {
...
       }
     }
-    return key;
+    return true;
+  }
+
+  /**
+   * Unregisters the given key from the global key map.
+   *
+   * @param key th property to unregister
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/PropertyKey.java]: nit: th -> the

File: core/common/src/test/java/alluxio/conf/AlluxioPropertiesTest.java (hunk 14 above)
Comment (apc999) [reply core/common/src/test/java/alluxio/conf/AlluxioPropertiesTest.java]: done

**Pull request merged**